src/com/amazon/ionschema/internal/constraint/ValidValues.kt

```diff
@@ -55,14 +44,33 @@ internal class ValidValues(
                 null
             }
 
+    private fun isValidRange(ion: IonValue) = ion is IonList && !ion.isNullValue && ion.hasTypeAnnotation("range")
+
+    private val anyRangeValues = validValues?.any{ isValidRange(it) } ?: false
+
     init {
         if (validRange == null && validValues == null) {
             throw InvalidSchemaException("Invalid valid_values constraint: $ion")
         }
     }
 
+    // build range value from given ion value
+    private fun buildRange(ion: IonValue) =
+        if (ion is IonList && isValidRange(ion)) {
+            if (ion[0] is IonTimestamp || ion[1] is IonTimestamp) {
```

Comment:
That gets tricky with `min` and `max` since they are Ion symbols. I believe the check that the range values are the appropriate types (including the special cased `min` and `max`) is in the `RangeFactory` code. https://github.com/amzn/ion-schema-kotlin/blob/master/src/com/amazon/ionschema/internal/util/Range.kt